components/datetime/src/neo.rs

```diff
@@ -803,47 +806,148 @@ impl<R: NeoFormatterMarker> NeoFormatter<R> {
-            ExtractedDateTimeInput::extract_from(datetime)
-        };
+        if !datetime.is_any_calendar_kind(self.calendar.kind()) {
+            return Err(crate::MismatchedCalendarError {
+                this_kind: self.calendar.kind(),
+                date_kind: <R::D as DateMarkers>::DateInputMarker::from(datetime)
+                    .into()
+                    .and_then(|v| v.any_calendar_kind),
+            });
+        }
+        let datetime = ExtractedDateTimeInput::extract_from_neo_input(
+            <R::D as DateMarkers>::DateInputMarker::from(datetime).into(),
+            <R::D as DateMarkers>::WeekInputMarker::from(datetime).into(),
+            <R::T as TimeMarkers>::TimeInputMarker::from(datetime).into(),
+        );
         Ok(FormattedNeoDateTime {
             pattern: self.selection.select(&datetime),
             datetime,
             names: self.names.as_borrowed(),
         })
     }
 
-    /// Infallibly formats a datetime after first converting it
+    /// Formats a datetime after first converting it
     /// to the formatter's calendar.
+    ///
+    /// # Examples
+    ///
+    /// Mismatched calendars convert and format automatically:
+    ///
+    /// ```
+    /// use icu::calendar::Date;
+    /// use icu::datetime::neo::NeoFormatter;
+    /// use icu::datetime::neo_marker::NeoYearMonthDayMarker;
+    /// use icu::datetime::neo_skeleton::NeoSkeletonLength;
+    /// use icu::datetime::MismatchedCalendarError;
+    /// use icu::locid::locale;
+    /// use writeable::assert_try_writeable_eq;
+    ///
+    /// let formatter = NeoFormatter::<NeoYearMonthDayMarker>::try_new(
+    ///     &locale!("en-u-ca-hebrew").into(),
+    ///     NeoSkeletonLength::Long,
+    /// )
+    /// .unwrap();
+    ///
+    /// let date = Date::try_new_roc_date(113, 5, 8).unwrap();
+    ///
+    /// assert_try_writeable_eq!(
+    ///     formatter.convert_and_format(&date),
+    ///     "30 Nisan 5784"
+    /// );
+    /// ```
+    ///
+    /// A time cannot be passed into the formatter when a date is expected:
+    ///
+    /// ```compile_fail
+    /// use icu::calendar::Time;
+    /// use icu::datetime::neo::NeoFormatter;
+    /// use icu::datetime::neo_marker::NeoYearMonthDayMarker;
+    /// use icu::datetime::neo_skeleton::NeoSkeletonLength;
+    /// use icu::locid::locale;
+    ///
+    /// let formatter = NeoFormatter::<NeoYearMonthDayMarker>::try_new(
+    ///     &locale!("es-MX").into(),
+    ///     NeoSkeletonLength::Long,
+    /// )
+    /// .unwrap();
+    ///
+    /// // the trait `for<'a> FromInCalendar<&'a icu::icu_calendar::Time>`
+    /// // is not implemented for `NeoDateInputFields`
+    /// formatter.convert_and_format(&Time::try_new(0, 0, 0, 0).unwrap());
+    /// ```
     pub fn convert_and_format<T>(&self, datetime: &T) -> FormattedNeoDateTime
     where
-        T: DateTimeInput<Calendar = Iso>,
+        T: ?Sized,
+        for<'a> <R::D as DateMarkers>::DateInputMarker: FromInCalendar<&'a T>,
```

Comment:
hm. not a fan of introducing HRTB here

suggestion: can we move the lifetime into the trait method? and use another custom trait for the other method.